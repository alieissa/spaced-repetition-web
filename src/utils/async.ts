/** @format */
/**
 * Credit goes to https://medium.com/@fillopeter/pattern-matching-with-typescript-done-right-94049ddd671c
 * For the pattern matching code
 */
export type Untriggered = {
  type: 'Untriggered'
}
export function Untriggered(): Untriggered {
  return {
    type: 'Untriggered',
  }
}

const t: Optional<Answer, 'easiness' | 'quality' | 'interval'> = {
  id: '',
  content: 'fsdfds',
}
export type Loading<V> = {
  type: 'Loading'
  value: V
}
export function Loading<V>(v: V): Loading<V> {
  return {
    type: 'Loading',
    value: v,
  }
}

export type Failure<E> = {
  type: 'Failure'
  value: E
}
export function Failure<E>(e: E): Failure<E> {
  return {
    type: 'Failure',
    value: e,
  }
}

export type Success<V> = {
  type: 'Success'
  value: V
}
export function Success<V>(v: V): Success<V> {
  return {
    type: 'Success',
    value: v,
  }
}

export type Async<L = unknown, F = unknown, S = unknown> =
  | Untriggered
  | Loading<L>
  | Failure<F>
  | Success<S>

export type AsyncStatus<R, E, A> = Async<R, E, A>['type']

export type AsyncMap<R, E, A, U> = {
  [K in AsyncStatus<R, E, A>]: U extends { type: K } ? U : never
}

export type AsyncStatusMap<R, E, A> = AsyncMap<R, E, A, Async<R, E, A>>
export type AsyncPattern<T, R, E, A> = {
  [K in keyof AsyncStatusMap<R, E, A>]: (a: AsyncStatusMap<R, E, A>[K]) => T
}

export function match<R, E, A>(
  e: Async<R, E, A>,
): <T>(pattern: AsyncPattern<T, R, E, A>) => T {
  return <T>(pattern: AsyncPattern<T, R, E, A>) => pattern[e.type](e as any)
}
