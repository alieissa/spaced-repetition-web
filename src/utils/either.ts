/** @format */

/**
 * Credit goes to https://medium.com/@fillopeter/pattern-matching-with-typescript-done-right-94049ddd671c
 * For the pattern matching code
 */
export type Right<V> = Readonly<{
  type: 'Right'
  value: V
}>
export function Right<V>(v: V): Right<V> {
  return {
    type: 'Right',
    value: v,
  }
}

export type Left<E = Error> = Readonly<{
  type: 'Left'
  value: E
}>
export function Left<E>(e: E): Left<E> {
  return {
    type: 'Left',
    value: e,
  }
}

export type Either<L, R> = Left<L> | Right<R>

export type EitherType<L, R> = Either<L, R>['type']
export type EitherMap<L, R, U> = {
  [K in EitherType<L, R>]: U extends { type: K } ? U : never
}
export type EitherTypeMap<L, R> = EitherMap<L, R, Either<L, R>>
export type EitherPattern<T, L, R> = {
  [K in keyof EitherTypeMap<L, R>]: (either: EitherTypeMap<L, R>[K]) => T
}

export function match<L, R>(
  e: Either<L, R>,
): <T>(pattern: EitherPattern<T, L, R>) => T {
  return <T>(pattern: EitherPattern<T, L, R>) => pattern[e.type](e as any)
}
