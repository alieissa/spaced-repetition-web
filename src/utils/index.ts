/** @format */

import * as async from './async'
import * as either from './either'

export { async }
export { either }
