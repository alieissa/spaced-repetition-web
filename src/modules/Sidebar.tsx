/** @format */

import { useEffect } from 'react'
import { useNavigate } from 'react-router-dom'
import { Icon, Menu, MenuItem, Sidebar } from 'semantic-ui-react'
import { styles } from 'src/styles'
import { useLogout } from './auth/auth.hooks'

export default function SPSidebar() {
  const navigate = useNavigate()
  const [logoutStatus, logout] = useLogout()

  // Does not make sense to NOT allow a user to logout
  // so we navigate to login page regardless of status of
  // logout call. The auth token is removed immediately.
  useEffect(() => {
    if (logoutStatus.type === 'Untriggered') {
      return
    }
    navigate('/login')
  }, [logoutStatus.type])

  return (
    <Sidebar
      as={Menu}
      visible={true}
      icon="labeled"
      vertical
      className="bordered bs-none"
    >
      <MenuItem as="a" onClick={() => navigate('/')}>
        <Icon name="home" />
        Decks
      </MenuItem>
      <MenuItem
        data-testid="sidebar-logout-menu-item"
        as="a"
        // Override very specific semantic ui css rule
        style={{ ...styles.absolute, ...styles.b0 }}
        onClick={() => logout()}
      >
        <Icon name="log out" />
        Logout
      </MenuItem>
    </Sidebar>
  )
}
