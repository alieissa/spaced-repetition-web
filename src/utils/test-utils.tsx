/** @format */

import { combineReducers, configureStore } from '@reduxjs/toolkit'
import { render } from '@testing-library/react'
import React from 'react'
import { Provider } from 'react-redux'
import { RouterProvider, createMemoryRouter } from 'react-router-dom'
import decksReducer from 'src/modules/decks/decks.reducer'

// As a basic setup, import your same slice reducers
import { loginReducer } from 'src/modules/login'
import { signupReducer } from 'src/modules/signup'
import { verificationReducer } from 'src/modules/verification'

// This type interface extends the default options for render from RTL, as well
// as allows the user to specify other things such as initialState, store.
// interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
//   preloadedState?: PreloadedState<RootState>
//   store?: AppStore
// }

const reducers = combineReducers({
  login: loginReducer,
  signup: signupReducer,
  verification: verificationReducer,
  decks: decksReducer,
})

const withRedux = (ui: React.ReactElement) => {
  const store = configureStore({ reducer: reducers })
  return <Provider store={store}>{ui}</Provider>
}
const withRouter = (ui: React.ReactElement) => {
  const router = createMemoryRouter([{ element: ui, path: '/' }], {
    initialEntries: ['/'],
    initialIndex: 1,
  })

  return <RouterProvider router={router} />
}
export function renderWithProviders(ui: React.ReactElement) {
  // Return an object with the store and all of RTL's query functions
  return render(withRedux(withRouter(ui)))
}

export const flushPromises = () =>
  new Promise((resolve) => setTimeout(resolve, 0))
