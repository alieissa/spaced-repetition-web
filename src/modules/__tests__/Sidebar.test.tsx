/** @format */

import '@testing-library/jest-dom'
import { act, screen, waitFor } from '@testing-library/react'
import { RestContext, rest } from 'msw'
import { setupServer } from 'msw/node'
import { flushPromises, renderWithProviders } from 'src/utils/test-utils'
import Sidebar from '../Sidebar'
const mockNavigate = jest.fn(() => ({}))
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockNavigate,
}))

const logoutUrl = `${process.env.REACT_APP_API_ENDPOINT}/users/logout`

export const handlers = [
  rest.get(logoutUrl, (__, res, ctx) => {
    return res(ctx.status(200))
  }),
]

const server = setupServer(...handlers)
beforeAll(() => server.listen())
afterEach(() => server.resetHandlers())
afterAll(() => server.close())

describe('Sidebar', () => {
  describe('view', () => {
    it('should render correctly', async () => {
      const { asFragment } = renderWithProviders(<Sidebar />)
      await act(flushPromises)

      expect(asFragment()).toMatchSnapshot()
    })
  })

  describe('interaction', () => {
    it.each([
      ['loading', (ctx: RestContext) => [ctx.status(200), ctx.delay(100)]],
      ['failed', (ctx: RestContext) => [ctx.status(422)]],
      ['success', (ctx: RestContext) => [ctx.status(200)]],
    ])(
      'should navigate to login when logout call is %s',
      async (_, apiMockReturn) => {
        // Assemble
        server.use(
          rest.get(logoutUrl, (__, res, ctx) => res(...apiMockReturn(ctx))),
        )
        renderWithProviders(<Sidebar />)

        // Act
        const logoutMenuItem = screen.getByTestId('sidebar-logout-menu-item')
        await act(() => logoutMenuItem.click())

        // Assert
        await waitFor(() => expect(mockNavigate).toHaveBeenCalledWith('/login'))
      },
    )
  })
})
