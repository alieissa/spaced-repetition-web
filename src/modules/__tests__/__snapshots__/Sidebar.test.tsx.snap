// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidebar view should render correctly 1`] = `
<DocumentFragment>
  <div
    class="ui vertical labeled icon ui left visible sidebar bordered bs-none menu"
  >
    <a
      class="item"
    >
      <i
        aria-hidden="true"
        class="home icon"
      />
      Decks
    </a>
    <a
      class="item"
      data-testid="sidebar-logout-menu-item"
      style="position: absolute; bottom: 0px;"
    >
      <i
        aria-hidden="true"
        class="log out icon"
      />
      Logout
    </a>
  </div>
</DocumentFragment>
`;
